### ISSUE TITLE
ATOM server error. Mintscan API returns INVALID_TOKEN.

### ISSUE THREAD

**Issue by dizzib** [hodgerpodger/staketaxcsv#340]
This download was working a few months ago…

![IMG_0364](https://github.com/hodgerpodger/staketaxcsv/assets/1223150/aa735e5c-ba3e-4cb1-867e-0742fe55443c)

Possibly related, [mintscan “Try API”](https://docs.cosmostation.io/apis/reference/historical/account/account-transactions) gives me this error:

```
{
statusCode:401,
message:"INVALID_TOKEN",
error:"Unauthorized",
path:"/v1/cosmos/accounts/cosmos1w33k********************************96pyhse/transactions?take=20&fromDateTime=2016-01-01",
timestamp:"2024-03-20T10:41:54.723Z"
}
```

I paste the api-key as the bear token.

Thanks!

**Comment by hodgerpodger** [comment]
hm I can't identify this issue in the logs (typically I can).  might try again or send email to support@stake.tax with a bookmark url.  Not sure at the moment how to identify the issue.

**Comment by hodgerpodger** [comment]
Oh and the mintscan api invalid token i think is probably just a usage error, or you need to obtain a key from mintscan.

**Comment by dizzib** [comment]
Thanks.

Not sure if this helps, but I’m getting the following error when trying to download ATOM on my local box…

```
INFO:root:localconfig: {'__module__': 'staketaxcsv.atom.config_atom', 'start_date': None, 'end_date': None, 'limit': 5000, '__doc__': None, 'job': None, 'debug': False, 'koinlynullmap': None}
INFO:root:koinly NullMap: use_cache=False, json_path=/staketaxcsv/src/staketaxcsv/common/../_reports/koinly_null_map.json
INFO:root:Requesting url https://apis.mintscan.io/v1/cosmos/accounts/cosmos1w33k********************************96pyhse/transactions?take=20&fromDateTime=2016-01-01%2000%3A00%3A00 ...
Traceback (most recent call last):
  File "/staketaxcsv/src/staketaxcsv/report_atom.py", line 91, in <module>
    main()
  File "/staketaxcsv/src/staketaxcsv/report_atom.py", line 22, in main
    report_util.main_default(TICKER_ATOM)
  File "/staketaxcsv/src/staketaxcsv/common/report_util.py", line 24, in main_default
    run_report(ticker, wallet_address, export_format, txid, options)
  File "/staketaxcsv/src/staketaxcsv/common/report_util.py", line 38, in run_report
    staketaxcsv.api.csv(ticker, wallet_address, export_format, path, options)
  File "/staketaxcsv/src/staketaxcsv/api.py", line 110, in csv
    exporter = module.txhistory(wallet_address)
  File "/staketaxcsv/src/staketaxcsv/common/ibc/decorators.py", line 16, in wrapper
    result = func(*args, **kwargs)
  File "/staketaxcsv/src/staketaxcsv/report_atom.py", line 67, in txhistory
    count_pages = txdata.get_txs_pages_count(wallet_address, start_date, end_date)
  File "/staketaxcsv/src/staketaxcsv/common/ibc/tx_data.py", line 56, in get_txs_pages_count
    return api_mintscan_v1.get_txs_page_count(
  File "/staketaxcsv/src/staketaxcsv/common/ibc/api_mintscan_v1.py", line 178, in get_txs_page_count
    num_txs = min(total_txs, max_txs)
TypeError: '<' not supported between instances of 'int' and 'NoneType'
```

I suspect the mintscan API is returning 401 but [_query](https://github.com/hodgerpodger/staketaxcsv/blob/0b993475b7bb9fc94eae28f6a15fd3c48b26c0af/src/staketaxcsv/common/ibc/api_mintscan_v1.py#L48) is not raising an exception (like it does for 406).

**Comment by hodgerpodger** [comment]
Not really sure what setup results in this error message (though I have seen it during development before somewhere long past).  In any case, you need a mintscan api key now with the newer implementation to be set for STAKETAX_MINTSCAN_KEY for AKT/ARCH/ATOM/DYDX/EVMOS/INJ/JUNO/OSMO/STRD/TIA .

If you really want to get into testing without a mintscan key, you can replace _txdata() in the code using TxDataLcd instead of TxDataMintscan.  There may be a detail or two you need to iron out doing this but it's something that will work, given a working LCD node with data (which probably is the more challenging part).

Another alternative is to use report_cosmoplus.py, as is.

**Comment by hodgerpodger** [comment]
I used to have an implementation where the code was flexible between mintscan api or LCD node as data source, but decided to simplify and just hard-code mintscan api.  Perhaps I will bring back that sort of logic if desirable.

**Comment by hodgerpodger** [comment]
Going to close for brevity's sake but feel free to comment, as I'll still follow along.

**Comment by dizzib** [comment]
Thank you for your help.

The 401 was caused by an invalid API key which was in a PENDING state. I had to email them to get it manually activated, which took a couple of days.

It is now ACTIVE and I can download transactions, both from their “Try API” page and from a local staketaxcsv container.

![IMG_0365](https://github.com/hodgerpodger/staketaxcsv/assets/1223150/96a7f191-b7ea-4837-bfe6-20d1ad034aef)
